logstash-core/lib/logstash/instrument/metric_store.rb

```diff
@@ -28,49 +35,107 @@ def fetch_or_store(namespaces, key, default_value = nil)
+    #
+    # The returned hash will keep the same structure as it had in the `Concurrent::Map`
+    # but will be a normal ruby hash. This will allow the api to easily seriliaze the content
+    # of the map
     #
     # @param [Array] The path where values should be located
     # @return nil if the values are not found
+    def get_with_path(path)
+      key_paths = path.gsub(/^#{KEY_PATH_SEPARATOR}+/, "").split(KEY_PATH_SEPARATOR)
+      get(*key_paths)
+    end
+
+    # Use an array of symbols instead of path
     def get(*key_paths)
-      get_recursively(key_paths, @store)
+      # Normalize the symbols access
+      key_paths.map(&:to_sym)
+      new_hash = Hash.new({})
+
+      get_recursively(key_paths, @store, new_hash)
+
+      new_hash
     end
 
-    # Return all the individuals Metric
-    #
-    # @return [Array] An array of all metric transformed in `Logstash::Event`, or in case of passing a block it yields
-    # the expected value as other Enumerable implementations.
-    def each(&block)
-      data = each_recursively(@store).flatten
-      if block_given?
-        data.each(&block)
+    # Return all the individuals Metric,
+    # This call mimic a Enum's each if a block is provided
+    def each(path = nil, &block)
+      metrics = if path.nil?
+        get_all
       else
-        return data
+        transform_to_array(get_with_path(path))
       end
+
+      block_given? ? metrics.each(&block) : metrics
     end
+    alias_method :all, :each
 
     private
-    def get_recursively(key_paths, map)
-      key_candidate = key_paths.shift
+    def get_all
+      each_recursively(@store).flatten
+    end
+
+    def get_recursively(key_paths, map, new_hash)
+      key_candidates = extract_filter_keys(key_paths.shift)
 
-      if key_paths.empty?
-        return map[key_candidate]
-      else 
-        next_map = map[key_candidate]
+      key_candidates.each do |key_candidate|
+        raise MetricNotFound, "For path: #{key_candidate}" if map[key_candidate].nil?
 
-        if next_map.is_a?(Concurrent::Map)
-          return get_recursively(key_paths, next_map)
+        if key_paths.empty? # End of the user requested path
+          if map[key_candidate].is_a?(Concurrent::Map)
+            new_hash[key_candidate] = transform_to_hash(map[key_candidate])
+          else
+            new_hash[key_candidate] = map[key_candidate]
+          end
         else
-          return nil
+          if map[key_candidate].is_a?(Concurrent::Map)
+            new_hash[key_candidate] = get_recursively(key_paths, map[key_candidate], {})
+          else
+            new_hash[key_candidate] = map[key_candidate]
+          end
         end
       end
+      return new_hash
     end
 
+    def extract_filter_keys(key)
+      key.to_s.strip.split(FILTER_KEYS_SEPARATOR).map(&:to_sym)
+    end
+
+    def transform_to_array(map)
+      map.values.collect do |value|
```

Comment:
lets say that you have the following structure
```
{ :foo => { :bar = > { :zoo => 1, :hello => { :ninja => 2 } } }
```
If you do `each("foo/bar")` you will get [1, 2] since they are both under `foo.bar`